### PR TITLE
Create a helper trait for parsing protobuf enums.

### PR DESCRIPTION
**Pull request by Spawek**

### CODE REVIEW

File: src/session/parse_enum.rs [hunk 1]
```diff
@@ -0,0 +1,63 @@
+use crate::session::UnknownEnumValueError;
+
+pub trait ProtoEnum<Proto> {
+    // Returns a default value for given Protobuf definition.
+    fn default() -> Proto;
+
+    // Returns value of the enum or None if the input i32 does not describe any know enum value.
+    // ::prost::Enumeration cannot be used instead, because it's not a trait.
```
Comment (panhania) [inline src/session/parse_enum.rs:8]: Identifiers are not formatted properly.

File: src/session/parse_enum.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    fn from_i32(val: i32) -> Option<Proto>;
+}
+
+// Maps the raw integer value to enum value or returns ParseError when the value cannot be mapped.
```
Comment (panhania) [inline src/session/parse_enum.rs:12]: Ditto.

File: src/session/parse_enum.rs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+pub fn parse_enum<T: ProtoEnum<T>>(raw_enum_value: Option<i32>) -> Result<T, UnknownEnumValueError> {
```
Comment (panhania) [inline src/session/parse_enum.rs:13]: Line too long. You can use `where` clasue to make it fit.

File: src/session/parse_enum.rs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        fn default() -> TestEnum { TestEnum::Two }
+        fn from_i32(val: i32) -> Option<TestEnum> {
+            TestEnum::from_i32(val)
+        }
+    }
+
+    #[test]
+    fn parse_empty_enum_test() {
+        let empty: Option<i32> = None;
+        let parsed : TestEnum = parse_enum(empty).unwrap();
```
Comment (panhania) [inline src/session/parse_enum.rs:46]: There should be no space before colon. Also, I don't think you need any type annotations here (and in these tests in general).

File: src/session/parse_enum.rs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        assert_eq!(parsed, TestEnum::Two);
+    }
+
+    #[test]
+    fn parse_correct_enum_value_test() {
+        let parsed : TestEnum = parse_enum(Some(1)).unwrap();
+        assert_eq!(parsed, TestEnum::One);
```
Comment (panhania) [inline src/session/parse_enum.rs:53]: Why not just `assert_eq!(parse_enum(Some(1)), Some(TestEnum::One))`?

File: src/session/parse_enum.rs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        fn from_i32(val: i32) -> Option<TestEnum> {
+            TestEnum::from_i32(val)
+        }
+    }
+
+    #[test]
+    fn parse_empty_enum_test() {
+        let empty: Option<i32> = None;
+        let parsed : TestEnum = parse_enum(empty).unwrap();
+        assert_eq!(parsed, TestEnum::Two);
```
Comment (panhania) [inline src/session/parse_enum.rs:47]: Why not just `assert_eq!(parse_enum(None), Some(TestEnum::Two))`

File: src/session/parse_enum.rs [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    use super::*;
+
+    #[derive(Debug, PartialEq, ::prost::Enumeration)]
+    pub enum TestEnum {
+        One = 1,
+        Two = 2,
+    }
+
+    impl ProtoEnum<TestEnum> for TestEnum {
+        fn default() -> TestEnum { TestEnum::Two }
```
Comment (panhania) [inline src/session/parse_enum.rs:37]: I think it is not very intuitive that a value named `Two` is a default and makes the assertion below rather unobvious. I would use three variants: `Default = 0`, `One = 1` and `Two = 2`.

File: src/session/error.rs [hunk 8]
```diff
@@ -179,3 +185,42 @@ impl From<MissingFieldError> for ParseError {
...
+pub struct UnknownEnumValueError {
+    pub enum_name: &'static str,
+    pub value: i32
+}
+
+impl UnknownEnumValueError {
+
+    /// Creates a new error indicating that a proto enum has a value for which the definition
+    /// is not known.
+    pub fn new(enum_name: &'static str, value: i32) -> UnknownEnumValueError {
```
Comment (panhania) [inline src/session/error.rs:200]: In general, I don't think this constructor is useful. I think initializing this error always with the record syntax has better readability.

File: src/session/parse_enum.rs [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+// Maps the raw integer value to enum value or returns ParseError when the value cannot be mapped.
+pub fn parse_enum<T: ProtoEnum<T>>(raw_enum_value: Option<i32>) -> Result<T, UnknownEnumValueError> {
+    match raw_enum_value {
+        Some(int_value) => match T::from_i32(int_value) {
+            Some(parsed_value) => Ok(parsed_value),
+            None => Err(UnknownEnumValueError::new(
+                std::any::type_name::<T>(),
+                int_value,
+            )),
```
Comment (panhania) [inline src/session/parse_enum.rs:20]: This further proves my comment about using record syntax for initialization.

File: src/session/parse_enum.rs [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+    // Returns value of the enum or None if the input i32 does not describe any know enum value.
+    // ::prost::Enumeration cannot be used instead, because it's not a trait.
+    fn from_i32(val: i32) -> Option<Proto>;
+}
+
+// Maps the raw integer value to enum value or returns ParseError when the value cannot be mapped.
+pub fn parse_enum<T: ProtoEnum<T>>(raw_enum_value: Option<i32>) -> Result<T, UnknownEnumValueError> {
+    match raw_enum_value {
+        Some(int_value) => match T::from_i32(int_value) {
```
Comment (panhania) [inline src/session/parse_enum.rs:15]: You could use the `ok_or_else` combinator.

**Review by panhania — CHANGES_REQUESTED** [review]

File: src/session/error.rs [hunk 11]
```diff
@@ -179,3 +185,42 @@ impl From<MissingFieldError> for ParseError {
         ParseError::malformed(error)
     }
 }
+
+/// An error type for situations where proto enum has a value for which the definition is not known.
```
[the remark this answers is not in the document]
Comment (Spawek) [reply src/session/error.rs:189]: Fixed.

File: src/session/error.rs [hunk 12]
```diff
@@ -179,3 +185,42 @@ impl From<MissingFieldError> for ParseError {
...
+#[derive(Debug)]
+pub struct UnknownEnumValueError {
+    pub enum_name: &'static str,
```
[the remark this answers is not in the document]
Comment (Spawek) [reply src/session/error.rs:192]: Changed to "name".

File: src/session/error.rs [hunk 13]
```diff
@@ -179,3 +185,42 @@ impl From<MissingFieldError> for ParseError {
...
+    pub value: i32
```
[the remark this answers is not in the document]
Comment (Spawek) [reply src/session/error.rs:193]: Added docstrings.

[the remark this answers is not in the document]
Comment (Spawek) [reply src/session/error.rs:193]: Added .

File: src/session/error.rs [hunk 14]
```diff
@@ -179,3 +185,42 @@ impl From<MissingFieldError> for ParseError {
...
+}
+
+impl UnknownEnumValueError {
+
+    /// Creates a new error indicating that a proto enum has a value for which the definition
+    /// is not known.
+    pub fn new(enum_name: &'static str, value: i32) -> UnknownEnumValueError {
+        UnknownEnumValueError {
+            enum_name,
+            value
```
[the remark this answers is not in the document]
Comment (Spawek) [reply src/session/error.rs:203]: This code was removed.

File: src/session/error.rs (hunk 8 above)
Comment (Spawek) [reply src/session/error.rs:200]: Removed.

File: src/session/error.rs [hunk 15]
```diff
@@ -179,3 +185,42 @@ impl From<MissingFieldError> for ParseError {
...
+    }
+}
+
+impl std::error::Error for UnknownEnumValueError {
+
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        None
+    }
+}
+impl From<UnknownEnumValueError> for ParseError {
```
[the remark this answers is not in the document]
Comment (Spawek) [reply src/session/error.rs:221]: Added.

File: src/session/error.rs [hunk 16]
```diff
@@ -179,3 +185,42 @@ impl From<MissingFieldError> for ParseError {
...
+
+    fn from(error: UnknownEnumValueError) -> ParseError {
+        ParseError::UnknownEnumValue(error)
+    }
+}
```
[the remark this answers is not in the document]
Comment (Spawek) [reply src/session/error.rs:226]: Added.

File: src/session/parse_enum.rs (hunk 1 above)
Comment (Spawek) [reply src/session/parse_enum.rs:8]: Fixed.

File: src/session/parse_enum.rs (hunk 9 above)
Comment (Spawek) [reply src/session/parse_enum.rs:20]: Fixed.

File: src/session/parse_enum.rs (hunk 7 above)
Comment (Spawek) [reply src/session/parse_enum.rs:37]: Changed to:
```
        Default = 1,
        Two = 2,
        Three = 3,
```

I don't want to use 0 as any value in here (and definitely not as default value). I don't want to have this test passing when 0 is returned by accident (which wouldn't be strange when playing with enums).

File: src/session/error.rs [hunk 17]
```diff
@@ -89,6 +89,8 @@ impl From<ParseError> for Error {
 pub enum ParseError {
     /// An error occurred because the decoded proto message was malformed.
     Malformed(Box<dyn std::error::Error + Send + Sync>),
+    /// A protobuf had a value which is now known.
+    UnknownEnumValue(UnknownEnumValueError),
```
[the remark this answers is not in the document]
Comment (Spawek) [reply src/session/error.rs]: Removed.

File: src/session/parse_enum.rs [hunk 18]
```diff
@@ -0,0 +1,61 @@
+use crate::session::UnknownEnumValueError;
+
+pub trait ProtoEnum<Proto> {
+    /// Returns a default value for given Protobuf definition.
+    fn default() -> Proto;
```
Comment (panhania) [inline src/session/parse_enum.rs:5]: This is covered by the standard `Default` trait.

File: src/session/parse_enum.rs [hunk 19]
```diff
@@ -0,0 +1,61 @@
+use crate::session::UnknownEnumValueError;
+
+pub trait ProtoEnum<Proto> {
```
Comment (panhania) [inline src/session/parse_enum.rs:3]: I am not sold on this trait. What is the problem is it trying to solve?

Comment (Spawek) [reply src/session/parse_enum.rs:3]: You can see more context in the PR: https://github.com/google/rrg/pull/42. For multiple protobuf enums we need to:
- know the default value (which is not a property of enum itself, but of the structure which uses the enum),
- be able to parse it easily,
- be able to detect that the received value of the enum is not recognised and return an approrpiate error.

With example implementation:
```
impl ProtoEnum<ActionType> for ActionType {
    fn default() -> Self {
        FileFinderAction::default().action_type()
    }
    fn from_i32(val: i32) -> Option<Self> {
        ActionType::from_i32(val)
    }
}
```

You can parse the enum from proto just by:

```
parse_enum::<ActionType>(proto.action_type)?;
```

File: src/session/parse_enum.rs (hunk 18 above)
Comment (Spawek) [reply src/session/parse_enum.rs:5]: I think my previous answer covers this comment also.

File: src/session/parse_enum.rs (hunk 19 above)
Comment (Spawek) [reply src/session/parse_enum.rs:3]: Small thing to add: in most of cases you parse by just `parse_enum(proto.action_type)` - no type annotations are needed.

File: src/session/parse_enum.rs (hunk 2 above)
Comment (Spawek) [reply src/session/parse_enum.rs]: Fixed.

File: src/session/parse_enum.rs (hunk 3 above)
Comment (Spawek) [reply src/session/parse_enum.rs]: Fixed - however without `where` clause.

File: src/session/error.rs [hunk 20]
```diff
@@ -180,25 +180,35 @@ impl From<MissingFieldError> for ParseError {
     }
 }
 
-/// An error type for situations where a given proto value is not supported.
+/// An error type for situations where proto enum has a value for which
+/// the definition is not known.
 #[derive(Debug)]
-pub struct UnsupportedValueError<T> {
-    /// A name of the field the value belongs to.
+pub struct UnknownEnumValueError {
```
Comment (panhania) [inline src/session/error.rs:186]: I think `UnsupportedValueError` and `UnknownEnumValueError` are different, I don't think one should replace another.

File: src/session/parse_enum.rs (hunk 18 above)
Comment (panhania) [reply src/session/parse_enum.rs:5]: I don't think it does (if it does, then please, explain).

Comment (Spawek) [reply src/session/parse_enum.rs:5]: The default value of any proto enum must be 0 ([link](https://developers.google.com/protocol-buffers/docs/proto3#:~:text=For%20enums%2C%20the%20default%20value,exact%20value%20is%20language%2Ddependent.)) In our case we care about the value to which we default when given proto enum is not set.

File: src/session/error.rs (hunk 20 above)
Comment (Spawek) [reply src/session/error.rs:209]: Brought back UnsupportedValueError - it was removed by accident during a merge.

**Review by panhania — APPROVED** [review]

**Pull request merged**